src/Components/Web.JS/src/Services/NavigationEnhancement.ts

```diff
@@ -0,0 +1,188 @@
+ - You can navigate from [1] to [2] by setting up the interactive <Router/> not to know about any [Route] components in your MPA,
+   and so it will fall back on a full-page load to get from the SPA URLs to the MPA URLs.
+ - You can navigate from [2] to [1] in that it just works by default. A <Router/> can be added dynamically and will then take
+   over and disable [2].
+
+Note that we don't reference NavigationManager.ts from NavigationEnhancement.ts or vice-versa. This is to ensure we could produce
+different bundles that only contain minimal content.
+*/
+
+let currentEnhancedNavigationAbortController: AbortController | null;
+let onDocumentUpdatedCallback: Function = () => {};
+
+export function attachProgressivelyEnhancedNavigationListener(onDocumentUpdated: Function) {
+  onDocumentUpdatedCallback = onDocumentUpdated;
+  document.body.addEventListener('click', onBodyClicked);
+  window.addEventListener('popstate', onPopState);
+}
+
+export function detachProgressivelyEnhancedNavigationListener() {
+  document.body.removeEventListener('click', onBodyClicked);
+  window.removeEventListener('popstate', onPopState);
+}
+
+function onBodyClicked(event: MouseEvent) {
+  if (hasInteractiveRouter()) {
+    return;
+  }
+
+  handleClickForNavigationInterception(event, absoluteInternalHref => {
+    history.pushState(null, /* ignored title */ '', absoluteInternalHref);
+    performEnhancedPageLoad(absoluteInternalHref);
+  });
+}
+
+function onPopState(state: PopStateEvent) {
+  if (hasInteractiveRouter()) {
+    return;
+  }
+
+  performEnhancedPageLoad(location.href);
+}
+
+async function performEnhancedPageLoad(internalDestinationHref: string) {
+  // First, stop any preceding enhanced page load
+  currentEnhancedNavigationAbortController?.abort();
+
+  // Now request the new page via fetch, and a special header that tells the server we want it to inject
+  // framing boundaries to distinguish the initial document and each subsequent streaming SSR update.
+  currentEnhancedNavigationAbortController = new AbortController();
+  const abortSignal = currentEnhancedNavigationAbortController.signal;
+  const responsePromise = fetch(internalDestinationHref, {
+    signal: abortSignal,
+    headers: { 'blazor-enhanced-nav': 'on' },
+  });
+  await getResponsePartsWithFraming(responsePromise, abortSignal,
+    (response, initialContent) => {
+      if (response.headers.get('content-type')?.startsWith('text/html')) {
+        // For HTML responses, regardless of the status code, display it
+        const parsedHtml = new DOMParser().parseFromString(initialContent, 'text/html');
+        synchronizeDomContent(document, parsedHtml);
+      } else if ((response.status < 200 || response.status >= 300) && !initialContent) {
+        // For any non-success response that has no content at all, make up our own error UI
+        document.documentElement.innerHTML = `Error: ${response.status} ${response.statusText}`;
```

Comment:
I don't think you'll actually see the redirects here.

AFAIK, fetch automatically follows the redirects.